### ISSUE TITLE
Failed install: gem requires Ruby version ~> 2.6. The current ruby version is 2.6.0.

### ISSUE THREAD

**Issue by ioquatix** [rubygems/rubygems#2343]
```
x_x > gem install /Users/samuel/Documents/ioquatix/covered/pkg/covered-0.4.0.gem
ERROR:  Error installing /Users/samuel/Documents/ioquatix/covered/pkg/covered-0.4.0.gem:
	There are no versions of covered (= 0.4.0) compatible with your Ruby & RubyGems
	covered requires Ruby version ~> 2.6. The current ruby version is 2.6.0.
```

My current problem is that it seems like when I add

```
spec.required_ruby_version = '~> 2.6'
```

to my `gemspec`, I can't install it on `2.6.0.preview2`.

  - [x] Installing a library
  - [x] Publishing a library
  - [x] The command line `gem`

Here are my current environment details:

```
$ gem env version
3.0.0.beta1
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by segiddins** [comment]
That is because `"2.6.0.preview2" < "2.6.0`"

**Comment by ioquatix** [comment]
That makes sense, but then why does it print out `The current ruby version is 2.6.0.` and not `2.6.0.preview2`?

I tried setting it to ">= 2.6.0.preview2" and it still didn't work. But, let me try again.

**Comment by ioquatix** [comment]
So...

```
^_^ > rake install
covered 0.4.0 built to pkg/covered-0.4.0.gem.
rake aborted!
Couldn't install gem, run `gem install /Users/samuel/Documents/ioquatix/covered/pkg/covered-0.4.0.gem' for more detailed output
/Users/samuel/.rvm/gems/ruby-2.6.0-preview2/gems/bundler-1.16.2/lib/bundler/gem_helper.rb:89:in `install_gem'
/Users/samuel/.rvm/gems/ruby-2.6.0-preview2/gems/bundler-1.16.2/lib/bundler/gem_helper.rb:46:in `block in install'
/Users/samuel/.rvm/gems/ruby-2.6.0-preview2/bin/ruby_executable_hooks:15:in `eval'
/Users/samuel/.rvm/gems/ruby-2.6.0-preview2/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => install
(See full trace by running task with --trace)
x_x > gem install /Users/samuel/Documents/ioquatix/covered/pkg/covered-0.4.0.gem
ERROR:  Error installing /Users/samuel/Documents/ioquatix/covered/pkg/covered-0.4.0.gem:
	There are no versions of covered (= 0.4.0) compatible with your Ruby & RubyGems
	covered requires Ruby version >= 2.6.0.preview2. The current ruby version is 2.6.0.
```

**Comment by segiddins** [comment]
Can you share the output of `gem env` ? Thanks!

**Comment by segiddins** [comment]
Ah that's because we intentionally use different versions for the check and the message -- https://github.com/rubygems/rubygems/blob/master/lib/rubygems/installer.rb#L634 -- check uses the actual version, and the message uses the API version

**Comment by ioquatix** [comment]
```
RubyGems Environment:
  - RUBYGEMS VERSION: 3.0.0.beta1
  - RUBY VERSION: 2.6.0 (2018-05-31 patchlevel -1) [x86_64-darwin17]
  - INSTALLATION DIRECTORY: /Users/samuel/.rvm/gems/ruby-2.6.0-preview2
  - USER INSTALLATION DIRECTORY: /Users/samuel/.gem/ruby/2.6.0
  - RUBY EXECUTABLE: /Users/samuel/.rvm/rubies/ruby-2.6.0-preview2/bin/ruby
  - GIT EXECUTABLE: /opt/local/bin/git
  - EXECUTABLE DIRECTORY: /Users/samuel/.rvm/gems/ruby-2.6.0-preview2/bin
  - SPEC CACHE DIRECTORY: /Users/samuel/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /Users/samuel/.rvm/rubies/ruby-2.6.0-preview2/etc
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-darwin-17
  - GEM PATHS:
     - /Users/samuel/.rvm/gems/ruby-2.6.0-preview2
     - /Users/samuel/.rvm/gems/ruby-2.6.0-preview2@global
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /Users/samuel/.rvm/gems/ruby-2.6.0-preview2/bin
     - /Users/samuel/.rvm/gems/ruby-2.6.0-preview2@global/bin
     - /Users/samuel/.rvm/rubies/ruby-2.6.0-preview2/bin
     - /Users/samuel/.rvm/bin
     - /opt/local/lib/postgresql96/bin
     - /opt/local/bin
     - /opt/local/sbin
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/sbin
     - /sbin
     - /opt/X11/bin
```

**Comment by ioquatix** [comment]
Do you want me to submit a PR to update the method you mentioned to be a bit clearer?

**Comment by segiddins** [comment]
I think that'd be helpful!

**Comment by MSP-Greg** [comment]
@segiddins

I'm using trunk, not a preview.  So to force only a version of trunk, one would need `'~> 2.6.0.dev'`?

**Comment by segiddins** [comment]
or `>= 2.6.a` or something like that

**Comment by ioquatix** [comment]
So, you are saying if I do `>= 2.6.a` it will work?

**Comment by MSP-Greg** [comment]
Interesting issue in terms of a 'trunk' build or a 'preview' build.  I assume you built from a 'preview' tarball?

I think the main difference between a 'trunk' build and a 'preview' build of the same svn is the version.h file, which changes `RUBY_PATCHLEVEL_STR` from 'dev' to 'previewX'  (included because some may be aware of this, others not...)

Using `~> 2.6.0dev` or `>= 2.6.0dev` works with a straight trunk build.

**Comment by deivid-rodriguez** [comment]
I had a look at this.

It happens because `rubygems` [assumes](https://github.com/rubygems/rubygems/blob/d9bbed7d35fe6fbcd955cf299f56bd8953ec96d1/lib/rubygems.rb#L966-L967) only prerelease versions define `RUBY_REVISION`, but it seems like all ruby versions actually define that (it's the revision part in the `ruby --version` output, for example, "ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]").

So the comparison done is `"2.6.0.dev.63539" >= "2.6.0.preview2"`, which fails.

The only soluton I've found is to manually parse `RUBY_DESCRIPTION` since `RUBY_PATCHLEVEL_STR` does not seem exposed. I can open a PR with this solution.

**Comment by ioquatix** [comment]
I tried setting to `~> 2.6.a` and it worked.

**Comment by bronzdoc** [comment]
@ioquatix is this still an issue for you?

**Comment by ioquatix** [comment]
Thanks!!
